pkg/omni/parser/parser_test.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2018 Vulcanize
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parser_test
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/vulcanize/vulcanizedb/examples/constants"
+	"github.com/vulcanize/vulcanizedb/pkg/geth"
+	"github.com/vulcanize/vulcanizedb/pkg/omni/parser"
+)
+
+var _ = Describe("Parser Test", func() {
+
+	var p parser.Parser
+	var err error
+
+	BeforeEach(func() {
+		p = parser.NewParser("")
+	})
+
+	It("Fetches and parses abi using contract address", func() {
+		contractAddr := "0x89d24a6b4ccb1b6faa2625fe562bdd9a23260359" // dai contract address
+		err = p.Parse(contractAddr)
+		Expect(err).ToNot(HaveOccurred())
+
+		expectedAbi := constants.DaiAbiString
+		Expect(p.Abi()).To(Equal(expectedAbi))
+
+		expectedParsedAbi, err := geth.ParseAbi(expectedAbi)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(p.ParsedAbi()).To(Equal(expectedParsedAbi))
+	})
+
+	It("Returns parsed methods and events", func() {
+		contractAddr := "0x89d24a6b4ccb1b6faa2625fe562bdd9a23260359"
+		err = p.Parse(contractAddr)
+		Expect(err).ToNot(HaveOccurred())
+
+		methods := p.GetMethods()
+		events := p.GetEvents()
+
+		_, ok := methods["totalSupply"]
+		Expect(ok).To(Equal(true))
+
+		_, ok = events["Transfer"]
+		Expect(ok).To(Equal(true))
+	})
+
+	It("Fails with a normal, non-contract, account address", func() {
+		addr := "0xAb2A8F7cB56D9EC65573BA1bE0f92Fa2Ff7dd165"
+		err = p.Parse(addr)
+		Expect(err).To(HaveOccurred())
+	})
+
```

Comment:
Wondering if we may want to have tests to document:

(1) What happens when the contract includes events above and beyond those defined on the erc20 spec. This may not be an issue for the parser but curious how it affects the larger system
(2) What happens when the contract doesn't have an abi defined on Etherscan (but is an erc20 contract)

I don't necessarily expect these would assert anything other than an error occurs at this point, but think it'd be good to document and modify as we augment the feature set

Reply:
If things are working properly #1 should not be an issue.
Number 2 is definitely an issue and the parser will throw an error. In general we need to find a way to get away from relying on Etherscan. One thing we could do for contracts whose ABIs are completely unavailable is assume a standard ERC20 ABI and try moving forward with that.

Reply:
Yeah makes sense. Long term I think it'd be cool to try and extract the contract code from the byte code accompanying the contract deploy (with something like [Rattle](https://github.com/trailofbits/rattle)) and then parse the abi out of that, but that's definitely not in scope for this PR. Mainly just thinking of tests as documentation for where we're at in the roadmap.

Reply:
I am skeptical that it will be possible to decompile byte code with enough fidelity to get an ABI back out but will look more into Rattle. It would be really awesome if we can.

Reply:
That's fair. I imagine this is a pretty ambitious ask. There was an interesting [talk](https://guidebook.com/guide/117233/event/21956134/) at Devcon from the developer of [Octopus](https://github.com/quoscient/octopus), and it seems like decompiling to that degree is increasingly feasible - but I wouldn't necessarily expect everything to just work for every contract out of the box.

Reply:
Thanks for those links! I will look into this extensively.